fastwarc/fastwarc/cli.py

```diff
@@ -167,13 +169,87 @@ def check(infile, decompress_alg, verify_payloads, quiet, output):
                     click.echo(rec)
                 sys.exit(1)
 
+@main.command()
+@click.argument('infile', type=click.Path(dir_okay=False, exists=True))
+@click.argument('offset', type=click.INT)
+@click.option('--payload', is_flag=True,
+              help='output only record payload (transfer and/or content encoding are preserved')
+@click.option('--headers', is_flag=True, help='output only record (and HTTP) headers')
+def extract(infile, offset, payload, headers):
+    """Extract WARC record by offset.
+    """
+    compl_record = not (payload or headers)
+    with open(infile, 'rb') as stream:
+        stream.seek(offset)
+        record = next(iter(ArchiveIterator(stream)))
+        if headers or compl_record:
+            ostream = PythonIOStreamAdapter(sys.stdout.buffer)
+            record.headers.write(ostream)
+            sys.stdout.buffer.write(b'\r\n')
+            if record.is_http:
+                record.http_headers.write(ostream)
+                sys.stdout.buffer.write(b'\r\n')
+        if payload or compl_record:
+            buf = record.reader.read(4096)
+            while buf:
+                sys.stdout.buffer.write(buf)
+                buf = record.reader.read(4096)
```

Comment:
`WarcRecord.write()` adds an additional double CRLF at the end: one to terminate the record, one to insert a newline before the next record. I think there should be at least one extra CRLF at the end here as well (primarily to prevent weird output effects if the website itself has no trailing newlines).

For writing a complete record, you can also use `WarcRecord.write(ostream)` directly.

Reply:
Ok. There should be no byte extra in case you want to save binary payload to a file - PDF, image, etc. If the entire record is extracted it should be a valid WARC record, so that users can do tricks such as `for offset in 1234 2345; do fastwarc extract big.warc.gz $offset | gzip; done >small.warc.gz`. I need to double-check whether these use case work.

Reply:
Yeah, that could indeed be an edge case. Not adding newlines, on the other hand, would produce output different from what the original WARC file was. Concatenating multiple outputs of `fastwarc extract` wouldn't result in the exact same byte data as the input, since the newlines between the record are missing (although you'd probably use `fastwarc recompress` for usecases like that).